False positive `consider-using-dict-comprehension` when creating a dict using a list of tuple where key AND value vary depending on the same condition

### Bug description

The following code snippet raises R1717:

```python
dict([(k, v) if v > 0 else (k, 0) for k, v in stuff.items()])
```

Both of these contain an `IfExp`, which I don't think can be converted into a dict comprehension. Please correct me if I'm wrong.

### Configuration

_No response_

### Command used

```shell
pylint mytest.py
```


### Pylint output

```shell
************* Module mytest
mytest.py:1:0: R1717: Consider using a dictionary comprehension (consider-using-dict-comprehension)
```


### Expected behavior

R1717 to not show up

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.6-dev0
Python 3.10.0 (default, Oct 13 2021, 06:45:00) [Clang 13.0.0 (clang-1300.0.29.3)]
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/.copyrite_aliases
+++ b/.copyrite_aliases
@@ -108,5 +108,10 @@
     ],
     "authoritative_mail": "bot@noreply.github.com",
     "name": "bot"
+  },
+  {
+    "mails": ["tushar.sadhwani000@gmail.com", "tushar@deepsource.io"],
+    "authoritative_mail": "tushar.sadhwani000@gmail.com",
+    "name": "Tushar Sadhwani"
   }
 ]

--- a/ChangeLog
+++ b/ChangeLog
@@ -11,6 +11,11 @@ Release date: TBA
 ..
   Put new features here and also in 'doc/whatsnew/2.13.rst'
 
+* Fixed false positive ``consider-using-dict-comprehension`` when creating a dict
+  using a list of tuples where key AND value vary depending on the same condition.
+
+  Closes #5588
+
 * ``used-before-assignment`` now considers that assignments in a try block
   may not have occurred when the except or finally blocks are executed.
 

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -32,6 +32,11 @@ Extensions
 Other Changes
 =============
 
+* Fixed false positive ``consider-using-dict-comprehension`` when creating a dict
+  using a list of tuples where key AND value vary depending on the same condition.
+
+  Closes #5588
+
 * By default, pylint does no longer take files starting with ``.#`` into account. Those are
   considered `emacs file locks`_. This behavior can be reverted by redefining the
   ``ignore-patterns`` option.

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -952,7 +952,7 @@ class RefactoringChecker(checkers.BaseTokenChecker):
                     key2, value2 = element.orelse.elts
                     if (
                         key1.as_string() != key2.as_string()
-                        and value1.as_string != value2.as_string()
+                        and value1.as_string() != value2.as_string()
                     ):
                         return
 

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -940,7 +940,7 @@ class RefactoringChecker(checkers.BaseTokenChecker):
                     return
 
                 # If we have an `IfExp` here where both the key AND value
-                # are different, then don't raise the issue.
+                # are different, then don't raise the issue. See #5588
                 if (
                     isinstance(element, nodes.IfExp)
                     and isinstance(element.body, (nodes.Tuple, nodes.List))

--- a/pylint/checkers/refactoring/refactoring_checker.py
+++ b/pylint/checkers/refactoring/refactoring_checker.py
@@ -934,9 +934,28 @@ class RefactoringChecker(checkers.BaseTokenChecker):
             and node.args
             and isinstance(node.args[0], nodes.ListComp)
         ):
-            if node.func.name == "dict" and not isinstance(
-                node.args[0].elt, nodes.Call
-            ):
+            if node.func.name == "dict":
+                element = node.args[0].elt
+                if isinstance(element, nodes.Call):
+                    return
+
+                # If we have an `IfExp` here where both the key AND value
+                # are different, then don't raise the issue.
+                if (
+                    isinstance(element, nodes.IfExp)
+                    and isinstance(element.body, (nodes.Tuple, nodes.List))
+                    and len(element.body.elts) == 2
+                    and isinstance(element.orelse, (nodes.Tuple, nodes.List))
+                    and len(element.orelse.elts) == 2
+                ):
+                    key1, value1 = element.body.elts
+                    key2, value2 = element.orelse.elts
+                    if (
+                        key1.as_string() != key2.as_string()
+                        and value1.as_string != value2.as_string()
+                    ):
+                        return
+
                 message_name = "consider-using-dict-comprehension"
                 self.add_message(message_name, node=node)
             elif node.func.name == "set":

--- a/tests/functional/c/consider/consider_using_dict_comprehension.py
+++ b/tests/functional/c/consider/consider_using_dict_comprehension.py
@@ -10,6 +10,7 @@ dict([(number, number*2) for number in numbers])  # [consider-using-dict-compreh
 
 stuff = {1: 10, 2: -20}
 dict([(k, v) if v > 0 else (k, 0) for k, v in stuff.items()])  # [consider-using-dict-comprehension]
+dict([(k, v) if v > 0 else (k*2, v) for k, v in stuff.items()])  # [consider-using-dict-comprehension]
 dict([(k, v) if v > 0 else (k * 2, 0) for k, v in stuff.items()])
 
 # Cannot emit as this cannot be written as a comprehension

--- a/tests/functional/c/consider/consider_using_dict_comprehension.py
+++ b/tests/functional/c/consider/consider_using_dict_comprehension.py
@@ -8,5 +8,9 @@ dict([])
 
 dict([(number, number*2) for number in numbers])  # [consider-using-dict-comprehension]
 
+stuff = {1: 10, 2: -20}
+dict([(k, v) if v > 0 else (k, 0) for k, v in stuff.items()])  # [consider-using-dict-comprehension]
+dict([(k, v) if v > 0 else (k * 2, 0) for k, v in stuff.items()])
+
 # Cannot emit as this cannot be written as a comprehension
 dict([value.split("=") for value in ["a=b", "c=d"]])

--- a/tests/functional/c/consider/consider_using_dict_comprehension.py
+++ b/tests/functional/c/consider/consider_using_dict_comprehension.py
@@ -1,4 +1,4 @@
-# pylint: disable=missing-docstring, invalid-name, use-dict-literal
+# pylint: disable=missing-docstring, invalid-name, use-dict-literal, line-too-long
 
 numbers = [1, 2, 3, 4, 5, 6]
 

--- a/tests/functional/c/consider/consider_using_dict_comprehension.txt
+++ b/tests/functional/c/consider/consider_using_dict_comprehension.txt
@@ -1,2 +1,3 @@
 consider-using-dict-comprehension:9:0:9:48::Consider using a dictionary comprehension:UNDEFINED
 consider-using-dict-comprehension:12:0:12:61::Consider using a dictionary comprehension:UNDEFINED
+consider-using-dict-comprehension:13:0:13:63::Consider using a dictionary comprehension:UNDEFINED

--- a/tests/functional/c/consider/consider_using_dict_comprehension.txt
+++ b/tests/functional/c/consider/consider_using_dict_comprehension.txt
@@ -1,1 +1,2 @@
 consider-using-dict-comprehension:9:0:9:48::Consider using a dictionary comprehension:UNDEFINED
+consider-using-dict-comprehension:12:0:12:61::Consider using a dictionary comprehension:UNDEFINED

